ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -127,6 +199,95 @@ def find_all_files(self, *args, **kwargs):
 
         Equivalent to ``list(self.find_files(...))``
         """
+
+        # NOTE: This code gets complex with the REGEX.
```

Comment:
I gather you extended `find_all_files` instead of the `find_files` it delegates to because of issues with `return`/`yield`? Because implemented like this, you'd only benefit from the cache when doing the former, while we generally use the latter.

Reply:
OK, so the cache as it is can only speed up lookup by `ID` and `fileGrp`. So:

* find_files(ID='...') -> cache
* find_files(fileGrp='...') -> cache
* find_files(ID='...', fileGrp='...') -> cache

In this case, we can yield those results.

If there are more kwargs to `find_files`, we can still use the cache for a list of candidates and then use the existing logic to drill down those candidates accordingly

> Then the `yield` returns all matches inside the `self._cache_flag` condition block and enters the `for cand in self._tree ...` loop to return 5 more files. 10 in total.

But why can we not deduplicate by `ID`? Or rather, why go through the XML if we have the cached information?

On a general note, have you considered [memoizing](https://docs.python.org/3/library/functools.html#functools.cache) the `find_files` method, so that calling `find_files` with the same arguments instantly returns the cached results?

Reply:
On another general note: In what circumstances would one want to disable the caching? I.e. do we need `_cache_flag`?

Reply:
> If there are more kwargs to `find_files`, we can still use the cache for a list of candidates and then use the existing logic to drill down those candidates accordingly

See d60c8d06e for what I mean with that.